### ISSUE TITLE
Misleading error message on traits after #8229

### ISSUE THREAD

**Issue by HollayHorvath** [rust-lang/rust-clippy#8306]
### Summary

I have a simple trait that extends the iterators (just like `itertools` does). This trait has `Iterator` bound, but my function also has `Iterator` in its `Self` argument (just a more precise one).

Before #8229 it was fine, now it hits `trait_duplication_in_bounds` (which is great!) but the error message is misleading and points to the function's `where`-clause, however removing that bound does not solve this issue.

### Reproducer

I tried this code:

```rust
pub trait Iter: Iterator {
    fn into_group_btreemap<K, V>(self) -> BTreeMap<K, Vec<V>>
    where
        Self: Iterator<Item = (K, V)> + Sized,
        K: Ord + Eq,
    {
        into_group_btreemap(self)
    }
}
```

I expected to see this happen:
The lint should show that the trait bound is superfluous:
```
3 |     pub trait Iter: Iterator {
  |                     ^^^^^^^^
```

Instead, this happened:
The lint shows the real bound:
```
6 |         Self: Iterator<Item = (K, V)> + Sized,
  |               ^^^^^^^^^^^^^^^^^^^^^^^
```

### Version

```text
rustc 1.60.0-nightly (bd3cb5256 2022-01-16)
binary: rustc
commit-hash: bd3cb52565faab2755ff1bdb54d88bc91f47b4b9
commit-date: 2022-01-16
host: x86_64-unknown-linux-gnu
release: 1.60.0-nightly
LLVM version: 13.0.0
```

**Comment by Jarcho** [comment]
Clippy should definitely not be suggesting to remove the super trait as it's not superfluous. It adds a constraint to the type implementing `Iter` which would otherwise not exist. The where clause is only restricting the default function implementation so other implementations of the trait are only bound by them if they do not provide an implementation. Any type constrained by `Iter` will be able to make use of any of `Iterator`'s items. e.g.

```rust
fn f<I: Iter>(mut i: I) -> Option<I::Item>  {
    i.next()
}
```

Removing the `Iterator` bound on the trait would make the previous code fail to compile.

With that said this is a false positive. The bound `Self: Iterator<Item = (K, V)>` can not be expressed without duplicating the bound on the trait.
